posts/2018-08-16-well-typed-heap.lhs

```diff
@@ -0,0 +1,1100 @@
+
+![](/images/draft-07.png)
+
+We [merge](#merge) the remainder heap from step 1 together with the new heap we
+made out of the children of the removed tree:
+
+![](/images/draft-08.png)
+
+The above merge requires carrying twice.
+
+Taking apart a single tree
+--------------------------
+
+We will start by implementing step 2 of the algorithm above since it is a bit
+easier.  In this step, we are taking all children from a tree and turning that
+into a new heap.
+
+We need to keep all our invariants intact, and in this case this means tracking
+them in the type system.  A tree of `o` has `2ᵒ` elements.  If we remove the
+root, we have `o` children trees with `2ᵒ - 1 ` elements in total.  Every child
+becomes a tree in the new heap.  This means that the heap contains `o` full
+trees, and its shape will be written as `o` "1"s.  This matches our math: if you
+write `o` "1"s, you get the binary notation of `2ᵒ - 1`.
+
+Visually:
+
+![A tree of order 3 results in a heap with a "11" shape](/images/draft-07.png)
+
+We introduce a type family for computing `n` "1"s:
+
+> type family Ones (n :: Nat) :: Binary where
+>     Ones 'Zero     = 'BEnd
+>     Ones ('Succ n) = 'B1 (Ones n)
+
+We will use a helper function `childrenToForest_go` to maintain some invariants.
+The wrapper `childrenToForest` is trivially defined but its type tells us a
+whole deal:
+
+> childrenToForest
+>     :: SNat n
+>     -> Children n a
+>     -> Forest 'Zero (Ones n) a
+> childrenToForest nnat children =
+>     childrenToForest_go SZero nnat FEnd children
+
+The tricky bit is that the list of trees in `Children` has them in descending
+order, and we want them in ascending order in `Forest`.  This means we will
+have to reverse the list.
+
+We can reverse a list easily using an accumulator in Haskell.  In order to
+maintain the type invariants at every step, we will increase the size of the
+accumulator as we descrease the size of the children.  This can be captured
+by requiring that their sum remains equal (`m ~ NAdd x n`).
+
+> childrenToForest_go
+>     :: m ~ NAdd x n
+>     => SNat x
+>     -> SNat n
+>     -> Forest n (Ones x) a
+>     -> Children n a
+>     -> Forest 'Zero (Ones m) a
+
+> childrenToForest_go xnat _snat@SZero acc CZero =
+
+I will not always go into detail on how the lemmas apply but let's do it here
+nonetheless.
+
+For the base case, we simply want to return our accumulator.  However, our
+accumulator has the type `Forest n (Ones x)` and we expect something of the type
+`Forest n (Ones m)`.  Furthermore, we know that:
+
+~~~~~~
+n ~ 'Zero, m ~ NAdd x n
+⊢ m ~ NAdd x 'Zero
+~~~~~~
+
+We need to prove that `x ~ m` in order to do the cast from `Forest n (Ones x)`
+to `Forest n (Ones m)`.
+
+We can do so by applying `lemma1` to `x` (the latter represented here by
+`xnat`).  This gives us `lemma1 xnat :: NAdd x 'Zero :~: n`.  Combining this
+with what we already knew:
+
+~~~~~~
+m ~ NAdd x 'Zero, NAdd x 'Zero ~ n
+⊢ m ~ x
```

Comment:
It looks a bit weird here but further down we get some long ones, which I need to line break.  I figured it would be best to do it consistently.  Maybe it looks better if I indent the assumptions with 2 spaces, I'll play around with it.

Reply:
maybe you can write it in a natural deduction style, with a line instead of a turnstile?